### PR TITLE
[v_array] Give v_array constructor, destructor, copy and move special functions

### PR DESCRIPTION
**Pull request by jackgerrits**
This will not pass CI, the purpose is to review small changes visually until we get to a point where we are back to working end to end.

### CODE REVIEW

File: vowpalwabbit/v_array.h [hunk 1]
```diff
@@ -39,25 +54,70 @@ struct v_array
...
+  inline const_iterator cbegin() const { return _begin; }
+  inline const_iterator cend() const { return _end; }
+
+  v_array() noexcept : _begin(nullptr), _end(nullptr), end_array(nullptr), erase_count(0) {}
+  ~v_array() { delete_v_array(); }
+
+  v_array(v_array<T>&& other) noexcept
+  {
+    erase_count = 0;
+    _begin = nullptr;
```
Comment (peterychang) [inline vowpalwabbit/v_array.h:81]: this might leak memory. Do we care?

File: vowpalwabbit/v_array.h [hunk 2]
```diff
@@ -31,6 +31,21 @@ const size_t erase_point = ~((1u << 10u) - 1u);
 template <class T>
 struct v_array
 {
+private:
+  void delete_v_array()
+  {
+    if (_begin != nullptr)
+    {
+      for (iterator item = _begin; item != _end; ++item) { item->~T(); }
```
Comment (peterychang) [inline vowpalwabbit/v_array.h:39]: The rest of v_array assumes it only contains trivial types, do we want to maintain that assumption (for now)? We can add non-trivial type support in one shot later

File: vowpalwabbit/v_array.h [hunk 3]
```diff
@@ -39,25 +54,70 @@ struct v_array
...
+    // TODO this should use the other version when T is trivially copyable and this otherwise.
+    copy_array_no_memcpy(*this, other);
+  }
+
+  v_array<T>& operator=(const v_array<T>& other)
+  {
+    if (this == &other) return *this;
+
+    delete_v_array();
+    copy_array_no_memcpy(*this, other);
```
Comment (olgavrou) [inline vowpalwabbit/v_array.h:117]: does the above TODO apply here too?

File: vowpalwabbit/v_array.h (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/v_array.h:81]: This is the move constructor, so there is not object that we could be leaking.

File: vowpalwabbit/v_array.h (hunk 2 above)
Comment (jackgerrits) [reply vowpalwabbit/v_array.h:39]: There are places where it does not contain trivial types at the moment, such as decision_scores. So I think we need this

File: vowpalwabbit/v_array.h (hunk 3 above)
Comment (jackgerrits) [reply vowpalwabbit/v_array.h:117]: Yes it does, I'll add it in.

File: vowpalwabbit/v_array.h [hunk 4]
```diff
@@ -39,25 +54,70 @@ struct v_array
...
+
+    std::swap(_begin, other._begin);
+    std::swap(_end, other._end);
+    std::swap(end_array, other.end_array);
+    std::swap(erase_count, other.erase_count);
+  }
+
+  v_array<T>& operator=(v_array<T>&& other) noexcept
+  {
+    delete_v_array();
```
Comment (olgavrou) [inline vowpalwabbit/v_array.h:93]: shouldn't we check for
`if (this == &other)` here too?

Comment (jackgerrits) [reply vowpalwabbit/v_array.h:93]: I believe the answer to this is no https://stackoverflow.com/questions/9322174/move-assignment-operator-and-if-this-rhs

**Review by olgavrou — APPROVED** [review]

**Pull request merged**